Hayeong

duol9 opened the pull request:
Add Api for Review (Create Review, Create ReviewCommnet, Get Review, Delete Review, Delete ReviewComment)

duol9 committed:
✨feat: add DTOs for Review functionality

 - Add ReviewRequestDto with toReviewEntity method
 - Add ReviewResponseDto with fromReviewEntity method

duol9 committed:
✨feat: add API for create reviews

- Validate order status to ensure the order is delivered before allowing review creation
- Convert ReviewRequestDto to Review entity and save it to the database

duol9 committed:
✨feat: add user and store references to Review entity

- Add User reference to Review for user-based queries
- Add Store reference to Review for store-based queries
- Enable more flexible query capabilities for reviews

duol9 committed:
✨feat: add API for creating review comments

- Validate store ownership before allowing comment creation
- Save review comment to the database and return response DTO

duol9 committed:
✨feat: add method to fetch review by ID with exception handling

- Add findReviewById method to handle cases where the review is not found
- Throw NotFoundException with appropriate error code if review does not exist

duol9 committed:
✨feat: add validation for ReviewRequestDto

- Add @NotNull validation for orderId to ensure it is required
- Add @Range and @NotNull validations for rating to enforce value range and presence
- Improve input validation for review creation

duol9 committed:
✨feat: add API for fetch store reviews

- Add API to fetch reviews for a specific store
- Include associated comments for each review

duol9 committed:
✨feat: add sorting reviews

- Implement sorting reviews by createdAt DESC
- Implement sorting review comments by createdAt ASC

duol9 committed:
✨feat: add API for deleting reviews

- Add API to delete reviews by setting isDeleted to true
- Validate review ownership before allowing deletion

duol9 committed:
✨feat: add API for deleting review comments

- Add API to delete review comments by setting isDeleted to true
- Validate review comment ownership before allowing deletion

duol9 committed:
♻️refactor: remove explicit isDeleted condition in queries

- Configure global filtering to automatically exclude deleted reviews
- Remove redundant isDeleted = false condition from review queries

duol9 committed:
✨feat: add rating range filter for review queries

- Implement dynamic filtering for reviews based on rating range
- Add validation for minimum and maximum rating values
- Ensure proper pagination and sorting in review queries

duol9 committed:
✨feat: add response codes for review and store errors

- Add specific error codes for review-related actions (e.g., not allowed, not found, access denied)
- Include error code for invalid rating range
- Add error code for store not found
- Improve error handling consistency across the application

duol9 committed:
💄style: apply code formatting for better readability

- Reformatted code to improve consistency and readability
- No functional changes made

duol9 committed:
♻️refactor: Migrate review fetching to QueryDSL

- Replaced the existing JPQL implementation with QueryDSL for improved readability and maintainability.
- Integrated `findAllReviewsWithComments` method using QueryDSL to fetch reviews and their associated comments in a single query.
- Removed the previous JPQL-based repository method as it is no longer needed.
- Enhanced performance by optimizing the query structure with `leftJoin` and grouping logic.

duol9 committed:
Merge branch 'dev' into hayeong

freedrawing commented on src/main/java/com/threemeals/delivery/domain/order/exception/OrderNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+import com.threemeals.delivery.domain.common.exception.AccessDeniedException;
+
+public class OrderNotFoundException extends AccessDeniedException {
+	public OrderNotFoundException() {
```
NotFound면 `extends NotFoundException`이 낫지 않을까요??

freedrawing commented on src/main/java/com/threemeals/delivery/domain/review/repository/ReviewRepositoryImpl.java:
```diff
@@ -0,0 +1,76 @@
+
+	@Override
+	public Page<GetReviewResponseDto> findAllReviewsWithComments(Long storeId, int minRating, int maxRating, Pageable pageable) {
+		QReview review = QReview.review;
+		QReviewComment reviewComment = QReviewComment.reviewComment;
```
static import 하면 조금 더 깔끔해질 거 같습니다!!

freedrawing commented on src/main/java/com/threemeals/delivery/domain/review/repository/ReviewRepositoryImpl.java:
```diff
@@ -0,0 +1,76 @@
+		QReview review = QReview.review;
+		QReviewComment reviewComment = QReviewComment.reviewComment;
+		
+		List<Tuple> joinComments = queryFactory
+			.select(review, reviewComment)
+			.from(review)
+			.leftJoin(reviewComment).on(reviewComment.review.eq(review))
+			.where(
+				review.store.id.eq(storeId),
+				review.rating.between(minRating, maxRating)
+			)
+			.offset(pageable.getOffset())
+			.limit(pageable.getPageSize())
+			.fetch();
```
`isDeleted = false` 로직 하나 있으면 좋을지도!!

freedrawing commented on src/main/java/com/threemeals/delivery/domain/review/repository/ReviewRepositoryImpl.java:
```diff
@@ -0,0 +1,76 @@
+				groupedComments.getOrDefault(tuple.get(review), Collections.emptyList())
+			))
+			.toList();
+		return PageableExecutionUtils.getPage(reviews, pageable, joinComments::size);
```
총 데이터 개수를 가져와야 할 것 같습니다!! 
예시 (틀릴 수도 있음)
```java
JpaQuery<Long> totalCount = queryFactory
        .select(review.count())
        .from(review)
        .where(
            review.store.id.eq(storeId),
            review.rating.between(minRating, maxRating)
        )
        
        .
        .
        .

return getPage(review, pageable, totalCount::fetchOne);        
```

freedrawing reviewed (commented):
Good!!

duol9 committed:
♻️refactor: Update OrderNotFoundException to extend NotFoundException

- Changed `OrderNotFoundException` to extend `NotFoundException` instead of `AccessDeniedException` for improved semantic accuracy.
- Updated the error code in the constructor to `ErrorCode.ORDER_NOT_FOUND` to better reflect the exception’s purpose.

duol9 committed:
♻️refactor: Add isDeleted filter in query and optimize totalCount for large datasets

- Replaced `isDeleted` filtering logic with an explicit condition in the query for better flexibility and control.
- Added a separate `queryCount` query to handle pagination efficiently, ensuring compatibility with large datasets.
- Improved overall query performance and maintainability.
